t/fixtures/proxy.lua

```diff
@@ -0,0 +1,222 @@
+        end
+
+        send(upstream, str .. cr_lf)
+
+        if length > 0 then
+            send(upstream, sock:receive(length) .. cr_lf)
+        end
+
+        sock:receive(2) -- last cr_lf
+    until length == 0
+
+    send(upstream, cr_lf)
+end
+
+local function _forward_body(sock, upstream, length)
+    ngx.log(ngx.DEBUG, 'reading ', length, ' bytes')
+    send(upstream, sock:receive(length))
+end
+
+local noop = function() end
+
+local function forward_http_stream(sock, upstream)
+    local read_header = sock:receiveuntil(cr_lf, inclusive)
+    local body_length, header_line
+    local body_reader = noop
+
+    repeat
+        header_line = read_header()
+        send(upstream, header_line)
+
+        local header = ngx_re_match(header_line, [[(?<name>[^:\s]+):\s*(?<value>.+)\r\n$]], "oj")
+
+        if header and str_lower(header.name) == 'content-length' then
+            body_length = tonumber(header.value)
+            body_reader = _forward_body
+        end
+
+        if header and str_lower(header.name) == 'transfer-encoding' and str_lower(header.value) == 'chunked' then
+            body_reader = _forward_chunked_body
+        end
+    until header_line == cr_lf
+
+    body_reader(sock, upstream, body_length)
+end
+
+local function proxy_http_request(sock, method, url, http_version, upstream)
+    local uri = assert(http.parse_uri(nil, url))
+    local port = uri[3]
+    local host = uri[2]
+    local ok, err = upstream:connect(host, port)
+
+    if ok then
+        ngx.log(ngx.DEBUG, 'connected to upstream ', host, ':', port)
+        ngx.ctx.upstream = upstream
+        ngx.ctx.upstream_host = host
+        ngx.ctx.upstream_port = port
+    else
+        ngx.log(ngx.ERR, 'failed to connect to ', host, ':', port, ' err: ', err)
+        return nil, err
+    end
+
+    local req_line = format('%s %s %s%s', method, uri[4] or '/', http_version, cr_lf)
+
+    send(upstream, req_line)
+
+    forward_http_stream(sock, upstream)
+    forward_http_stream(upstream, sock)
+    ngx.log(ngx.DEBUG, 'finished forwarding http stream')
+
+    upstream:setkeepalive()
+end
+
+local function proxy_connect_request(sock, connect, upstream)
+    local ok, err = upstream:connect(connect.host, connect.port)
+
+    if ok then
+        ngx.log(ngx.DEBUG, 'connected to upstream ', connect.host, ':', connect.port)
+        ngx.ctx.upstream = upstream
+        ngx.ctx.upstream_host = connect.host
+        ngx.ctx.upstream_port = connect.port
+    else
+        ngx.log(ngx.ERR, 'failed to connect to ', connect.host, ':', connect.port, ' err: ', err)
+        return nil, err
+    end
+
+    local header_line
+    local read_request_header = sock:receiveuntil(cr_lf, inclusive)
+
+    repeat
+        header_line = read_request_header()
+        ngx.log(ngx.DEBUG, 'got header line: ', header_line)
+    until header_line == cr_lf
+
+    send(sock, http_ok)
+    send(sock, cr_lf)
+
+    return pipe_stream(sock, upstream)
+end
+
+return function()
```

Comment:
Understanding what this function does is not complicated. However, as it's the only method exposed by the module, I'd document it. I think it will help communicate the purpose of this module.